client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,165 @@
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerInstance( handler );
+	}
+
+	syncHandlerInstance( handler ) {
+		const self = this;
+
+		return function( params, wrappedResponseCallback ) {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return self.reqHandler( params, wrappedResponseCallback );
+			}
+
+			const clonedParams = Object.assign( {}, params );
+			const { path } = clonedParams;
+
+			// response sent flag
+			let responseSent = false;
+
+			// generate an unique resource key
+			const key = self.generateKey( params );
+
+			debug( 'starting to get resource ...' );
+
+			self.retrieveRecord( key, function( err, localRecord ) {
+				if ( err ) {
+					// @TODO improve error handling here
+					console.error( err );
+				}
+
+				// let's be optimistic
+				if ( localRecord ) {
+					responseSent = true;
+
+					debug( '%o stored(%o). Let\'s be optimistic ...\n', path, localRecord );
+					wrappedResponseCallback( null, localRecord.body );
+				} else {
+					debug( 'No data for %o\n', path );
+				}
+
+				// * background sync process *
+				debug( 'requesting %o to WP.com', path );
+				handler( params, ( resErr, resData ) => {
+					if ( resErr ) {
+						if ( responseSent ) {
+							return;
+						}
+
+						return wrappedResponseCallback( resErr );
+					}
+
+					debug( 'WP.com response %o -> %o', path, resData );
+
+					const isPostRequest = clonedParams &&
+						clonedParams.method &&
+						'post' === clonedParams.method.toLowerCase();
+
+					if ( ! isPostRequest ) {
+						if ( responseSent ) {
+							debug( 'data is already stored. Overwriting ...' );
+						}
+
+						// remove some response fields before to store
+						delete resData._headers;
+
+						let storingData = {
+							__sync: {
+								key,
+								synced: new Date().toString(),
+								syncing: false
+							},
+							body: resData,
+							params: clonedParams
+						};
+
+						self.storeRecord( key, storingData );
+					} else {
+						debug( 'skip - non sync request: [%o] %o - %o\n', path, params, resData );
+					}
+
+					if ( ! responseSent ) {
+						wrappedResponseCallback( err, resData );
+					}
+				} );
+			} );
+		};
+	}
+
+	/**
+	 * Generate a key from the given param object
+	 *
+	 * @param {Object} params - request parameters
+	 * @param {Boolean} hashCodification - codificate key when it's true
+	 * @return {String} request key
+	 */
+	generateKey( params, hashCodification = true ) {
+		var key = params.apiVersion || '';
+		key += '-' + params.method;
+		key += '-' + params.path;
+
+		if ( params.query ) {
+			key += '-' + params.query;
+		}
+
+		if ( hashCodification ) {
+			key = new Hashes.SHA1().hex( key );
+		}
+
+		debug( 'key: %o', key );
+		return key;
+	}
+
+	retrieveRecord( key, fn = () => {} ) {
```

Comment:
Is the named `fn` helpful here?
I would recommend either a helpful name or anonymous function